gobblin-runtime/src/main/java/gobblin/runtime/SafeDatasetCommit.java

```diff
@@ -0,0 +1,275 @@
+        || this.jobContext.getJobCommitPolicy() == JobCommitPolicy.COMMIT_SUCCESSFUL_TASKS
+        || (this.jobContext.getJobCommitPolicy() == JobCommitPolicy.COMMIT_ON_FULL_SUCCESS
+        && datasetState.getState() == JobState.RunningState.SUCCESSFUL);
+  }
+
+  @SuppressWarnings("unchecked")
+  private Optional<CommitSequence.Builder> generateCommitSequenceBuilder(JobState.DatasetState datasetState) {
+    try (Closer closer = Closer.create()) {
+      Class<? extends CommitSequencePublisher> dataPublisherClass =
+          (Class<? extends CommitSequencePublisher>) Class.forName(datasetState
+              .getProp(ConfigurationKeys.DATA_PUBLISHER_TYPE, ConfigurationKeys.DEFAULT_DATA_PUBLISHER_TYPE));
+      CommitSequencePublisher publisher =
+          (CommitSequencePublisher) closer.register(DataPublisher.getInstance(dataPublisherClass, this.jobContext.getJobState()));
+      publisher.publish(datasetState.getTaskStates());
+      return publisher.getCommitSequenceBuilder();
+    } catch (Throwable t) {
+      log.error("Failed to generate commit sequence", t);
+      setTaskFailureException(datasetState.getTaskStates(), t);
+      throw Throwables.propagate(t);
+    }
+  }
+
+  void checkForUnpublishedWUHandling(String datasetUrn, JobState.DatasetState datasetState,
+      Class<? extends DataPublisher> dataPublisherClass, Closer closer)
+      throws ReflectiveOperationException, IOException {
+    if (UnpublishedHandling.class.isAssignableFrom(dataPublisherClass)) {
+      // pass in jobstate to retrieve properties
+      DataPublisher publisher = closer.register(DataPublisher.getInstance(dataPublisherClass, this.jobContext.getJobState()));
+      log.info(String.format("Calling publisher to handle unpublished work units for dataset %s of job %s.",
+          datasetUrn, this.jobContext.getJobId()));
+      ((UnpublishedHandling) publisher).handleUnpublishedWorkUnits(datasetState.getTaskStatesAsWorkUnitStates());
+    }
+  }
+
+  private void finalizeDatasetState(JobState.DatasetState datasetState, String datasetUrn) {
+    for (TaskState taskState : datasetState.getTaskStates()) {
+      // Backoff the actual high watermark to the low watermark for each task that has not been committed
+      if (taskState.getWorkingState() != WorkUnitState.WorkingState.COMMITTED) {
+        taskState.backoffActualHighWatermark();
+        if (this.jobContext.getJobCommitPolicy() == JobCommitPolicy.COMMIT_ON_FULL_SUCCESS) {
+          // Determine the final dataset state based on the task states (post commit) and the job commit policy.
+          // 1. If COMMIT_ON_FULL_SUCCESS is used, the processing of the dataset is considered failed if any
+          //    task for the dataset failed to be committed.
+          // 2. Otherwise, the processing of the dataset is considered successful even if some tasks for the
+          //    dataset failed to be committed.
+          datasetState.setState(JobState.RunningState.FAILED);
```

Comment:
I don't even think this is necessary as the  finalizeDatasetStateBeforeCommit() should have already set the dataset state.